File: chronicle-pc/pipeline/chronicle_pipeline/captions.py
"""Persist vision image captions under index/ (Mac-only, not synced)."""

from __future__ import annotations

import logging
from pathlib import Path
from typing import Any

from .paths import atomic_write_json, read_json

log = logging.getLogger("chronicle.captions")


def captions_path(root: Path) -> Path:
    return Path(root) / "index" / "image_captions.json"


def load_captions(root: Path) -> dict[str, str]:
    path = captions_path(root)
    if not path.is_file():
        return {}
    try:
        data = read_json(path)
    except (OSError, ValueError) as e:
        log.warning("Failed to load image captions %s: %s", path, e)
        return {}
    if not isinstance(data, dict):
        return {}
    captions = data.get("captions") if "captions" in data else data
    if not isinstance(captions, dict):
        return {}
    out: dict[str, str] = {}
    for k, v in captions.items():
        if isinstance(k, str) and isinstance(v, str):
            out[k] = v
    return out


def save_captions(root: Path, captions: dict[str, str]) -> Path:
    path = captions_path(root)
    cleaned = {k: v for k, v in captions.items() if isinstance(k, str) and isinstance(v, str)}
    payload: dict[str, Any] = {"version": 1, "captions": cleaned}
    atomic_write_json(path, payload)
    return path

File: chronicle-pc/server/src/captions.rs
//! Image caption cache (captions.py) — index/image_captions.json storage.

use std::collections::HashMap;
use std::path::Path;

use serde_json::Value;

use crate::errors::ChronicleError;
use crate::paths::atomic_write_json;

pub fn captions_path(root: &Path) -> std::path::PathBuf {
    root.join("index").join("image_captions.json")
}

/// load_captions: tolerant; accepts wrapped {"captions": {…}} or flat map;
/// keeps only str→str pairs.
pub fn load_captions(root: &Path) -> HashMap<String, String> {
    let path = captions_path(root);
    let raw = match std::fs::read_to_string(&path) {
        Ok(r) => r,
        Err(_) => return HashMap::new(),
    };
    let data: Value = match serde_json::from_str(&raw) {
        Ok(d) => d,
        Err(e) => {
            eprintln!(
                "WARNING chronicle.captions: Failed to load image captions {}: {}",
                path.display(),
                e
            );
            return HashMap::new();
        }
    };
    let map_ref = data
        .get("captions")
        .and_then(Value::as_object)
        .map(|m| m.clone())
        .or_else(|| data.as_object().cloned());
    let Some(obj) = map_ref else { return HashMap::new() };
    obj.into_iter()
        .filter_map(|(k, v)| v.as_str().map(|s| (k, s.to_string())))
        .collect()
}

pub fn save_captions(root: &Path, captions: &HashMap<String, String>) -> Result<(), ChronicleError> {
    let mut inner = serde_json::Map::new();
    for (k, v) in captions {
        inner.insert(k.clone(), Value::String(v.clone()));
    }
    let payload = serde_json::json!({"version": 1, "captions": inner});
    atomic_write_json(&captions_path(root), &payload)
}
